templates/vm.xml.j2

```diff
@@ -22,7 +22,7 @@
     <disk type='volume' device='{{ volume.device | default(libvirt_volume_default_device) }}'>
       <driver name='qemu' type='{{ volume.format | default(libvirt_volume_default_format) }}'/>
       <source pool='{{ volume.pool }}' volume='{{ volume.name }}'/>
-      <target dev='vd{{ 'abcdefghijklmnopqrstuvwxyz'[loop.index] }}'/>
+      <target dev='vd{{ 'abcdefghijklmnopqrstuvwxyz'[loop.index - 1] }}'/>
```

Comment:
Interesting. I'm a bit worried about changing this now though for idempotence. We could get odd behaviour if we run this role against an existing VM which starts at vdb.

Reply:
> We could get odd behaviour if we run this role against an existing VM which starts at vdb.

Ah, yes, fair point! But only if VMs are first destroyed and then re-created. 

`TASK [stackhpc.libvirt-vm : Ensure the VM is defined]` reports unchanged when I swich between `[loop.index - 1]` or `[loop.index]`, and so, it won't impact existing VMs. This does lead to the interesting question of not being able to change existing VM config, but the workaround is simply to destroy/remove and re-create the VMs since changing definitions ins't coded for in the role.

Unless a user has hardcoded something else/external to depend on /dev/vdb being the first device in the guest, I think you can safely accept the index off by one correction I added. However, should you prefer, I can revert that correction and leave the existing logic that skips a and starts at b (but I'd argue that's not what most people creating a VM expect). Arguably, best practise is to not use device names but rather labels or UUIDs, but often people might assume vda or sda is the first OS device.

Reply:
There was an attempt to support changing the XML config: https://github.com/stackhpc/ansible-role-libvirt-vm/pull/14. I think we got stuck, it needs to be revived.

I know people shouldn't rely on the device names, but I also know that they do anyway :)

Ok, let's make it work as expected (as you have done), and people can fix issues with device names as/when they come up.

Reply:
> There was an attempt to support changing the XML config: #14. I think we got stuck, it needs to be revived.

My view is that if changing XML is meant to have an effect, the upstream `virt` ansible module should be the place to support it, otherwise you'd end up wrapping/changing it's behaviour. 

## Don't define/undefine by default

#14 's suggested behaviour might make the role do more work when VMs are already defined and haven't changed. For example, when there's been no change to the original XML domain definition, undefining and redefining by default could break/risk idempotency. VMs get UUIDs, and on EL, SELinux levels for Multi-Level Security (MLS) and Multi-Category Security (MCS) that might keep changing every time the role is run.

## Determining if the definition has changed is non-trivial

The underlying issue is that libvirt modifies the original XML given when defining a VM do if one runs, for example, run `virsh define --file <domain>.xml`, start and stop the VM, and then `virsh dumpxml <domain>`. You'll notice it doesn't match! E.g. dumpxml has extra items and the XML won't match the original definition. The only way I foresee the upstream `virt` module managing to gracefully handle this idempotency problem is to run a careful sub-match of XML sub-elements of the current define given vs what the API returns, ignore the irrelevant XML from runtime, and then use the equivalent mechanism behind `virsh edit` to change VM settings only when something intentional has changed.

Reply:
You are quite right about detecting changes being non-trivial. Even if we only check for changes in the XML provided by the user, this would not allow any elements to be removed from the XML. Perhaps the only sane way is a specific argument to force an update of the definition.

Reply:
> specific argument to force an update of the definition.

Yip, wouldn't be the first time an ansible module or role took this approach.